util/src/main/java/io/zeebe/util/health/CriticalComponentsHealthMonitor.java

```diff
@@ -0,0 +1,141 @@
+  private final Map<String, HealthMonitorable> monitoredComponents = new HashMap<>();
+  private final Map<String, HealthStatus> componentHealth = new HashMap<>();
+  private volatile HealthStatus healthStatus = HealthStatus.UNHEALTHY;
+  private final ActorControl actor;
+  private final Logger log;
+  private FailureListener failureListener;
+
+  public CriticalComponentsHealthMonitor(final ActorControl actor, final Logger log) {
+    this.actor = actor;
+    this.log = log;
+  }
+
+  public void startMonitoring() {
+    actor.runAtFixedRate(HEALTH_MONITORING_PERIOD, this::updateHealth);
+  }
+
+  @Override
+  public void monitorComponent(final String componentName) {
+    actor.run(() -> componentHealth.put(componentName, HealthStatus.UNHEALTHY));
+  }
+
+  @Override
+  public void removeComponent(final String componentName) {
+    actor.run(
+        () -> {
+          monitoredComponents.remove(componentName);
+          componentHealth.remove(componentName);
+        });
+  }
+
+  @Override
+  public void registerComponent(final String componentName, final HealthMonitorable component) {
+    actor.run(
+        () -> {
+          monitoredComponents.put(componentName, component);
+          component.addFailureListener(new ComponentFailureListener(componentName));
+          componentHealth.put(componentName, component.getHealthStatus());
+          calculateHealth();
+        });
+  }
+
+  @Override
+  public HealthStatus getHealthStatus() {
+    return healthStatus;
+  }
+
+  @Override
+  public void addFailureListener(final FailureListener failureListener) {
+    actor.run(() -> this.failureListener = failureListener);
+  }
+
+  private void onComponentFailure(final String componentName) {
+    actor.run(
+        () -> {
+          log.error("{} failed, marking it as unhealthy", componentName);
+          componentHealth.put(componentName, HealthStatus.UNHEALTHY);
+          calculateHealth();
+        });
+  }
+
+  private void onComponentRecovered(final String componentName) {
+    actor.run(
+        () -> {
+          log.error("{} recovered, marking it as healthy", componentName);
+          componentHealth.put(componentName, HealthStatus.HEALTHY);
+        });
+  }
+
+  private void updateHealth() {
+    componentHealth
+        .keySet()
+        .forEach(component -> componentHealth.put(component, getHealth(component)));
+    calculateHealth();
+  }
+
+  private void calculateHealth() {
+    final var status =
+        componentHealth.containsValue(HealthStatus.UNHEALTHY)
+            ? HealthStatus.UNHEALTHY
+            : HealthStatus.HEALTHY;
+    final var previousStatus = healthStatus;
+    healthStatus = status;
+
+    if (failureListener != null && previousStatus != status) {
+      switch (status) {
+        case HEALTHY:
+          failureListener.onRecovered();
+          break;
+        case UNHEALTHY:
+          failureListener.onFailure();
+          break;
+        default:
```

Comment:
Here and above, I would add a warn logging - in the case we ever add more health statuses, then we'd immediately be aware of it.